Upcasting causes incompatible types in assignment with `TypedDict`

**Bug Report**

Type checking the code below causes `mypy` errors. In my opinion, everything assignable to type `T` should be still assignable to type `T | something`.

**To Reproduce**

[Playground](https://mypy-play.net/?mypy=latest&python=3.10&flags=strict&gist=fe3ea1dbc6216cf9557b9747c5946a49) with the following code:

```python
from typing import TypedDict


class AngleDict(TypedDict):
    yaw: float
    pitch: float
    roll: float
    

angles: dict[str, None] = {"yaw": None, "pitch": None, "roll": None}
angles2: AngleDict | dict[str, None] = {"yaw": None, "pitch": None, "roll": None}
```

**Expected Behavior**

No errors.

**Actual Behavior**

```
main.py:11: error: Incompatible types (expression has type "None", TypedDict item "yaw" has type "float")  [typeddict-item]
main.py:11: error: Incompatible types (expression has type "None", TypedDict item "pitch" has type "float")  [typeddict-item]
main.py:11: error: Incompatible types (expression has type "None", TypedDict item "roll" has type "float")  [typeddict-item]
Found 3 errors in 1 file (checked 1 source file)
```

**Your Environment**

- Mypy version used: master and 1.1.1
- Mypy command-line flags: nope
- Mypy configuration options from `mypy.ini` (and other config files): nope
- Python version used: 3.10, 3.11

--- a/mypy/checkexpr.py
+++ b/mypy/checkexpr.py
@@ -5350,9 +5350,9 @@ class ExpressionChecker(ExpressionVisitor[Type], ExpressionCheckerSharedApi):
         # an error, but returns the TypedDict type that matches the literal it found
         # that would cause a second error when that TypedDict type is returned upstream
         # to avoid the second error, we always return TypedDict type that was requested
-        typeddict_contexts = self.find_typeddict_context(self.type_context[-1], e)
+        typeddict_contexts, exhaustive = self.find_typeddict_context(self.type_context[-1], e)
         if typeddict_contexts:
-            if len(typeddict_contexts) == 1:
+            if len(typeddict_contexts) == 1 and exhaustive:
                 return self.check_typeddict_literal_in_context(e, typeddict_contexts[0])
             # Multiple items union, check if at least one of them matches cleanly.
             for typeddict_context in typeddict_contexts:
@@ -5363,7 +5363,8 @@ class ExpressionChecker(ExpressionVisitor[Type], ExpressionCheckerSharedApi):
                 self.chk.store_types(tmap)
                 return ret_type
             # No item matched without an error, so we can't unambiguously choose the item.
-            self.msg.typeddict_context_ambiguous(typeddict_contexts, e)
+            if exhaustive:
+                self.msg.typeddict_context_ambiguous(typeddict_contexts, e)
 
         # fast path attempt
         dt = self.fast_dict_type(e)
@@ -5425,22 +5426,29 @@ class ExpressionChecker(ExpressionVisitor[Type], ExpressionCheckerSharedApi):
 
     def find_typeddict_context(
         self, context: Type | None, dict_expr: DictExpr
-    ) -> list[TypedDictType]:
+    ) -> tuple[list[TypedDictType], bool]:
+        """Extract `TypedDict` members of the enclosing context.
+
+        Returns:
+            a 2-tuple, (found_candidates, is_exhaustive)
+        """
         context = get_proper_type(context)
         if isinstance(context, TypedDictType):
-            return [context]
+            return [context], True
         elif isinstance(context, UnionType):
             items = []
+            exhaustive = True
             for item in context.items:
-                item_contexts = self.find_typeddict_context(item, dict_expr)
+                item_contexts, item_exhaustive = self.find_typeddict_context(item, dict_expr)
                 for item_context in item_contexts:
                     if self.match_typeddict_call_with_dict(
                         item_context, dict_expr.items, dict_expr
                     ):
                         items.append(item_context)
-            return items
+                exhaustive = exhaustive and item_exhaustive
+            return items, exhaustive
         # No TypedDict type in context.
-        return []
+        return [], False
 
     def visit_lambda_expr(self, e: LambdaExpr) -> Type:
         """Type check lambda expression."""

--- a/test-data/unit/check-typeddict.test
+++ b/test-data/unit/check-typeddict.test
@@ -4297,9 +4297,13 @@ class TD(TypedDict):
     x: int
 
 x: Union[TD, dict[str, str]] = {"x": "foo"}
+y: Union[TD, dict[str, int]] = {"x": "foo"}  # E: Dict entry 0 has incompatible type "str": "str"; expected "str": "int"
 
-def accepts_td(d: Union[TD, dict[str, str]]) -> None: ...
+def ok(d: Union[TD, dict[str, str]]) -> None: ...
+ok({"x": "foo"})
 
-accepts_td({"x": "foo"})
+def bad(d: Union[TD, dict[str, int]]) -> None: ...
+bad({"x": "foo"})  # E: Dict entry 0 has incompatible type "str": "str"; expected "str": "int"
+
 [builtins fixtures/dict.pyi]
 [typing fixtures/typing-typeddict.pyi]

--- a/test-data/unit/check-typeddict.test
+++ b/test-data/unit/check-typeddict.test
@@ -4289,3 +4289,17 @@ inputs: Sequence[Component] = [{
 }]
 [builtins fixtures/dict.pyi]
 [typing fixtures/typing-typeddict.pyi]
+
+[case testTypedDictAssignableToWiderContext]
+from typing import TypedDict, Union
+
+class TD(TypedDict):
+    x: int
+
+x: Union[TD, dict[str, str]] = {"x": "foo"}
+
+def accepts_td(d: Union[TD, dict[str, str]]) -> None: ...
+
+accepts_td({"x": "foo"})
+[builtins fixtures/dict.pyi]
+[typing fixtures/typing-typeddict.pyi]

